Python: Copy sanitizes default modification

yoff opened the pull request:
Add ESSA definitions via copy as sanitiser.
Currently only tested on the FP snapshot.
I would write tests, but I first need to figure out why the unmodified tests fail...

yoff commented:
Jenkins seems to see the same as me. No new results, but many results duplicated.

yoff commented:
Apparently the configuration created the many duplicated results, and apparently it now works without it. Perhaps I fixed the sanitiser in the meantime.

yoff commented:
Ok, so tests pass now. I should look into a more robust detection of "value gets copied" I think, the current one only catches the posted FP.

yoff commented on python/ql/src/Functions/ModificationOfParameterWithDefault.ql:
```diff
@@ -93,7 +93,47 @@ class Mutation extends TaintSink {
+        or
+        value = Value::named("copy.deepcopy").getACall()
+        or
+        value.(AttrNode).getName() = "copy"
```
I am not sure this is needed when we have the case below.

tausbn replied:
Sounds to me like adding tests might supply the necessary certainty. :) 

As far as I can tell, this disjunct covers everything of the form `foo.copy`, whereas the next one covers `foo.copy()`. As such, I'm not sure this disjunct is useful.

yoff replied:
Well, it can take an awful lot of tests to distinguish equivalent things. I thought it better to get confirmation on the semantics :)

I also considered if the first form was in fact more general and therefor better (since it also seems to match the `foo.copy` part of `foo.copy()`), but having not read a bit about Python attributes, that would be a strange way to implement a copy functionality (so now you will tell me that of course someone did just that).

tausbn commented on python/ql/test/query-tests/Functions/general/functions_test.py:
```diff
@@ -198,3 +198,37 @@ def tuple_default(x=(1,2)):
 def safe_method(x=[]):
     return x.count(42)
+
+# Modification of parameter with default (sanitised)
```
Very minor nit: we aim to use US English in our documentation, so get your `z`s ready for writing `sanitization` etc. :slightly_smiling_face:

tausbn commented on python/ql/test/query-tests/Functions/general/functions_test.py:
```diff
@@ -198,3 +198,37 @@ def tuple_default(x=(1,2)):
+    x = x.copy().reverse()
+    return x.append(42)
+
+def tupple_copy_before_modify(x=[], y=[]):
```
Typo: `tuple`

tausbn commented on python/ql/src/Functions/ModificationOfParameterWithDefault.ql:
```diff
@@ -93,7 +93,47 @@ class Mutation extends TaintSink {
+        or
+        value.(AttrNode).getName() = "copy"
+        or
+        value.(CallNode).getNode().getFunc().(Attribute).getName() = "copy"
```
Going through the AST is unnecessary here. `CallNode`s have a `getFunction` method.

tausbn commented on python/ql/src/Functions/ModificationOfParameterWithDefault.ql:
```diff
@@ -93,7 +93,47 @@ class Mutation extends TaintSink {
 
-from TaintedPathSource src, TaintedPathSink sink
-where src.flowsTo(sink)
+/** copying prevents modification of the default value */
```
This could do with a bit more elaborate documentation. Something like
"By default, copying preserves most kinds of taint. However, for tracking mutable values, copying in fact _sanitizes_ the value being tracked, since mutating the copy does not affect the original."

tausbn commented on python/ql/test/query-tests/Functions/general/functions_test.py:
```diff
@@ -198,3 +198,37 @@ def tuple_default(x=(1,2)):
+    return x.append(42)
+
+def unsuccessfully_call_copy_before_modify(x=[]):
+    x = x.reverse().copy() # Unintentially modifies here
```
`x.reverse()` reverses `x` in-place and returns `None`, so this code does not work (since `None` has no `copy` method).

yoff commented on python/ql/src/Functions/ModificationOfParameterWithDefault.ql:
```diff
@@ -93,7 +93,47 @@ class Mutation extends TaintSink {
+        isCopied(def.(AssignmentDefinition).getValue())
+    }
+
+    predicate isCopied(ControlFlowNode value) {
```
It was meant to express that `value` is something being copied. Perhaps `creates_a_copy` is a bit clearer. I would like to avoid committing to a method of copy creation :)

yoff commented on python/ql/test/query-tests/Functions/general/functions_test.py:
```diff
@@ -198,3 +198,37 @@ def tuple_default(x=(1,2)):
+from copy import copy, deepcopy
+def copy_before_modify(x=[]):
+    x = copy(x)
+    return x.append(42)
```
I will change it. Our target is realistic code, so we should not test on weird code without good reason.

yoff commented:
Cool, I shall figure out how to run tests via `cpython` :)

tausbn commented:
Oh, I didn't notice your earlier comments about the duplicated test results. This is to be expected, due to the fact that you're now using a configuration. In the olden days, we didn't use configurations at all, but relied only on whatever classes you had imported for the given query. To make things a bit more similar to the other languages, we added the configurations. This left us with a bunch of queries that would yield no results, for lack of an appropriate configuration. Because of this, we have a `LegacyConfiguration` class that takes care of these configuration-less queries. Previously, it would only be active when no other configurations were present, but this turned out to have some unfortunate consequences, so eventually we just made it active all the time. This does mean that you see twice the edges and nodes you saw previously -- one comes from the configuration you provided, and the other from the legacy configuration.

yoff commented:
Thanks, Larsen explained this to me also, I guess I could have edited the previous posts to reflect my new state of less confusion.

yoff commented:
Rewrote the tests, I had also duplicated some previous ones. And now I can test that test code behaves as I expect inside VSCode :)
Also simplified the query as suggested.

yoff commented:
And it came to pass that Jenkins in its wisdom removed the false positive, thus building a better and brighter future, but leaving the developer rather confused.

yoff commented:
It is the alert
```
| functions_test.py:226:5:226:8 | Subscript | functions_test.py:224:51:224:51 | empty mutable value | functions_test.py:226:5:226:8 | empty mutable value | $@ flows to here and is mutated. | functions_test.py:224:51:224:51 | y | Default value |
```
which I get locally but not on Jenkins. It pertains to this function in `function_test.py`:
```
def tuple_unsuccessfully_copy_before_modify(x=[], y=[]):
    t = (x.copy(), y) # Whole tuple tainted here
    t[0].append(42) # FP here, since we taint both t[0] and t[1]
    return t[0]
```
Apparently, this FP does not occur on Jenkins.

yoff commented:
Aha, I can see that I have also received an email from Jenkins about a query change for `py/modification-of-default-value`

yoff commented:
After having pulled from `upstream/master`, restarted VSCode, and upgraded the CLI, I still see the FP locally :(

yoff commented:
Ok, I think I should write down our conclusions so far. This PR currently suffers from three separate issues:
 - **Too much flow**
In the program
```python
def tuple_unsuccessfully_copy_before_modify(x=[]):
    t = (copy(x), 3) # Tuple wrongly tainted here
    t[0].append(42) # FP here
    return t[0]
```
taint flows into the tuple `t` even though it should be sanitised by the call to `copy`.
 - **Too little flow**
In the program
```python
def tuple_copy_before_modify(x=[]):
    t = (x.copy(), 3) # Tuple not tainted here
    t[0].append(42)
    return t[0]
```
taint does not flow into the tuple `t`, but not because it is sanitised. Rather because taint flowing into tuples by the standard mechanism `from member` requires the flow to be between two `ControlFlowNode`s. However, `TaintTrackingNode`s are annotated `DataFlow::Node`s and a `DataFlow::Node` can be either a `ControlFlowNode` or an `EssaVariable`. As it happens, taint is flowing to `x.copy()` but only as represented by an `EssaVariable`, and so it does not flow further into the tuple.
 - **Py2/3 discrepancy**
The program in [Too much flow] does not exhibit too much flow when treated as being Python 2.

I hope that [Too much flow] can be handled by implementing further sanitisers. I Propose that [Too little flow] could be moved to a separate issue, perhaps the current behaviour is even desirable/intended. I think we must solve [Py2/3 discrepancy] before the present PR can be merged.

RasmusWL commented:
I'm confused about [Too little flow], it sounds more like a case of "correct result, but for wrong reasons"?

yoff commented:
It may be a case of me having untrained expectations of our data flow implementation. I would expect that if taint flows into a component of a tuple, then it flows, in sequence form, into the tuple.

RasmusWL commented:
> It may be a case of me having untrained expectations of our data flow implementation. I would expect that if taint flows into a component of a tuple, then it flows, in sequence form, into the tuple.

I'm not sure how to interpret this response :sweat_smile:

So just to be clear: the function `tuple_copy_before_modify` is safe, since it does not modify the default value. I would expect that `x.copy()` is not tainted, so `t = (x.copy(), 3)` should not make `t` a tainted tuple either. Is the current situation that:

1. `x.copy()` is not tainted, `t` is not tainted (all good :heavy_check_mark:)
2. `x.copy()` is tainted, but `t` is not tainted because of the SSA thing you mention (not good :x:)

yoff commented:
3. `x.copy()` the `ControlFlowNode` node is not tainted, but an `EssaVariable` with the same location is

yoff commented on python/ql/src/Functions/ModificationOfParameterWithDefault.ql:
```diff
@@ -100,19 +100,28 @@ class Mutation extends TaintSink {
-        (kind instanceof EmptyMutableValue
-        or
-        kind instanceof NonEmptyMutableValue) and
+    override predicate sanitizingDefinition(TaintKind taint, EssaDefinition def) {
```
It is possible that this is no longer needed..

yoff commented:
I commented out the function that behaves differently between py2/3 for now..

yoff commented:
Closing for now as this is getting out-dated.
